File: src/Template/Template.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace PatternCatalog
{
    class Template
    {
        public abstract class Workflow
        {
            public abstract void YouShouldDoThisFirst();
            public abstract void ThenYouShouldDoThis();
            public abstract void ThisWillBeDoneLast();

            // this is the 
            public void Process()
            {
                YouShouldDoThisFirst();
                ThenYouShouldDoThis();
                ThisWillBeDoneLast();
            }
        }

        public class Headhunter : Workflow
        {
            public override void YouShouldDoThisFirst()
            {
                Console.WriteLine("Lock the target");
            }

            public override void ThenYouShouldDoThis()
            {
                Console.WriteLine("Bait the line");
            }

            public override void ThisWillBeDoneLast()
            {
                Console.WriteLine("Spread the net");
            }
        }
    }
}

File: src/Template/template.rb
class Workflow
  def process
    you_should_do_this_first
    then_you_should_do_this
    this_will_be_done_last
  end
end

class Headhunter < Workflow

  def you_should_do_this_first
    puts("Lock the target")
  end

  def then_you_should_do_this
    puts("Bait the line")
  end

  def this_will_be_done_last
    puts("Spread the net")
  end

end

template = Headhunter.new()

template.process

# begin tests
describe Headhunter do 
 
  before do
    @headhunter = Headhunter.new()
  end

  subject { @headhunter }

  it { should respond_to(:you_should_do_this_first) }
  it { should respond_to(:then_you_should_do_this) }
  it { should respond_to(:this_will_be_done_last) }
 
  it "should call methods in order" do
    
    @headhunter.should_receive(:you_should_do_this_first).ordered
    @headhunter.should_receive(:then_you_should_do_this).ordered
    @headhunter.should_receive(:this_will_be_done_last).ordered

    @headhunter.process
  end
 
end